### PR TITLE
[feat] Add Refund logic in place

### PR DESCRIPTION
**Pull request by shekohex**
### Overview

The Refund process works like the following:

1. User requests some native tokens to be sent to the recipient along with the wrapped (or unwrapped) tokens.
2. The relayer validates this logic and sends the requested amount as a native token value to the smart contract.
3. we process the refund if and only if the value of the refund requested is greater than zero and equal to `msg.value` sent to the smart contract from the relayer.
4. Once processed, the value is sent to the recipient, if that failed we return it to the relayer or revert.

### Task Checklist

- [x] Initial Implementation
- [x] Documentation
- [x] Add Happy Path tests
- [x] Add negative tests.

### CODE REVIEW

**Comment by shekohex** [comment]
Hey @drewstone or @saiakilesh Could you help writing some tests for that? If not, could you please point me where to write these tests, and what is the workflow for running them?

**Comment by drewstone** [comment]
I would write them alongside the regular vanchor tests. You’ll want to
build the classes so that they have this logic of course.

You can create a separate file for the refund tests to keep it tidy. We
just need to:
1. Do a deposit
2. Relay a withdrawal with a non zero refund to a “relayer” (simulate this
ofc). The relayer will need to submit the withdrawal tx with a msg.value
that matches the refund. Then we should verify both the withdrawn token
balance of the recipient and the native balance of the recipient to show
the refund succeeded. Negative tests would be the opposite (messing with
values)

**Comment by saiakilesh** [comment]
@drewstone @shekohex Added some tests. The only change I made to Shady's contract work is that we should only refund during a withdraw (aka `extAmount < 0`).

**Comment by shekohex** [comment]
I'm working on fixing these failing tests right now, but for some reason, Locally only one test fails while on CI there are two of them!

File: packages/contracts/test/vanchor/vanchor.test.ts [hunk 1]
```diff
@@ -358,6 +359,140 @@ describe('VAnchor for 1 max edge', () => {
...
+      const ethBalanceAfter = await ethers.provider.getBalance(recipient);
+
+      assert.strictEqual(
+        ethBalanceAfter.sub(ethBalanceBefore).toString(),
+        BigNumber.from(1e4).toString()
+      );
+    });
+
+    it('should not refund upon deposit', async () => {
+      // Alice deposits into tornado pool
```
Comment (drewstone) [inline packages/contracts/test/vanchor/vanchor.test.ts:416]: Fix comments. Remove Tornado

**Review by drewstone — CHANGES_REQUESTED** [review]
LG overall, would be interested to hear you thoughts on tests.

File: packages/contracts/test/vanchor/vanchor.test.ts [hunk 2]
```diff
@@ -358,6 +359,140 @@ describe('VAnchor for 1 max edge', () => {
...
+
+      const ethBalanceAfter = await ethers.provider.getBalance(recipient);
+
+      assert.strictEqual(
+        ethBalanceAfter.sub(ethBalanceBefore).toString(),
+        BigNumber.from(0).toString()
+      );
+    });
+
+    it('should not refund upon internal transfer', async () => {
```
[the remark this answers is not in the document]
Comment (shekohex) [reply packages/contracts/test/vanchor/vanchor.test.ts:443]: I guess the main reason is that the `extAmount` here would be Zero. @saiakilesh any idea why we do not need to want to allow this?

[the remark this answers is not in the document]
Comment (saiakilesh) [reply packages/contracts/test/vanchor/vanchor.test.ts:443]: @drewstone 

This is specifically the case where `extAmount` is 0. So no funds are deposited/withdrawn, only commitments are added to the Merkle tree and nullifiers are stored on the smart contract. In this case, I thought there is no need for a refund?

**Comment by drewstone** [comment]
> @drewstone @shekohex Added some tests. The only change I made to Shady's contract work is that we should only refund during a withdraw (aka `extAmount < 0`).

Let's chat today about this, want to make sure the user story makes sense entirely.

**Pull request merged**